sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreCheckpointFormatV2Suite.scala

```diff
@@ -211,6 +221,323 @@ class RocksDBStateStoreCheckpointFormatV2Suite extends StreamTest
     }
   }
 
+  val changelogEnabled =
+    "spark.sql.streaming.stateStore.rocksdb.changelogCheckpointing.enabled" -> "true"
+  val changelogDisabled =
+    "spark.sql.streaming.stateStore.rocksdb.changelogCheckpointing.enabled" -> "false"
+  val ckptv1 = SQLConf.STATE_STORE_CHECKPOINT_FORMAT_VERSION.key -> "1"
+  val ckptv2 = SQLConf.STATE_STORE_CHECKPOINT_FORMAT_VERSION.key -> "2"
+
+  val testConfigSetups = Seq(
+    // Enable and disable changelog under ckpt v2
+    (Seq(changelogEnabled, ckptv2), Seq(changelogEnabled, ckptv2)),
+    (Seq(changelogDisabled, ckptv2), Seq(changelogDisabled, ckptv2))
+  )
+
+  testConfigSetups.foreach {
+    case (firstRunConfig, secondRunConfig) =>
+      testWithRocksDBStateStore("checkpointFormatVersion2 Backward Compatibility - simple agg - " +
+        s"first run: (changeLogEnabled, ckpt ver): " +
+        s"${firstRunConfig(0)._2}, ${firstRunConfig(1)._2}" +
+        s" - second run: ${secondRunConfig(0)._2}, ${secondRunConfig(1)._2}") {
```

Comment:
this is very hard to read imho. You have Seq of tuples of Seq.

The way I read it right now is that you run it first batch enabled ->  second batch enabled and first batch disabled -> second batch disabled.
Do you not need to test more combinations?

Reply:
Yes previously there were some v1 related test cases but now we focus on v2 only first. So the only combination I can think of is turning on / off changelog in v2. cc @siying

Reply:
I thought we decided that one time  V1 -> V2 migration should be supported and V2->V1 can be delayed to implement later. But we still should test it to make sure it throws an exception. I assume you've already had tests for both? It is a good point that, we might switch from (V1, changelogOn) -> (V2, changeLogOff) and vice versa.
I realized that (V1, changelogOn) -> (V2, changelogOn) is actually not supported now? Perhaps we should make sure that at least an exception is thrown to fail it.

I think even within V2, we might would like to cover the case such as changelogOff -> changelogOn -> changelogOff, ideally.

 I echo @brkyvz 's comment on using case class to improve readability.